juju/model.py

```diff
@@ -0,0 +1,167 @@
+        """
+        num_related = len(self.relations[relation_name])
+        if num_related > 1:
+            raise TooManyRelatedApps(relation_name, num_related, 1)
+        elif num_related == 0:
+            return None
+        else:
+            return self.relations[relation_name][0]
+
+
+class Application:
+    def __init__(self, name):
+        self.name = name
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+
+class Unit:
+    def __init__(self, name, app):
+        self.name = name
+        self.app = app
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+
+class RelationMap(Mapping):
+    """Map of relation names to lists of Relation instances."""
+    def __init__(self, relation_names, backend):
+        self._relation_names = relation_names
+        self._backend = backend
+
+    def __iter__(self):
+        return iter(self._relation_names)
+
+    def __len__(self):
+        return len(self._relation_names)
+
+    def __getitem__(self, relation_name):
+        return self._backend.get_relations(relation_name)
+
+
+class Relation:
+    def __init__(self, relation_name, relation_id, backend):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend
+        self.data = RelationData(self.relation_name, relation_id, self._backend)
+
+    @property
+    def apps(self):
+        return [unit.app for unit in self.units]
+
+    @property
+    def units(self):
+        return list(self.data.keys())
+
+
+class RelationData(Mapping):
+    def __init__(self, relation_name, relation_id, backend):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend
+        self._keys = self._backend.get_relation_members(self.relation_id)
+
+    def keys(self):
+        return self._keys
+
+    def __contains__(self, key):
+        return key in self._keys
+
+    def __iter__(self):
+        return iter(self._keys)
+
+    def __len__(self):
+        return len(self._keys)
+
+    def __getitem__(self, member):
+        if member not in self._keys:
+            raise KeyError(member)
+        return self._backend.relation_get(self.relation_id, member.name)
+
+
+class ModelError(Exception):
+    pass
+
+
+class TooManyRelatedApps(ModelError):
+    def __init__(self, relation_name, num_related, max_supported):
+        super().__init__(f'Too many remote applications on {relation_name} ({num_related} > {max_supported})')
+        self.relation_name = relation_name
+        self.num_related = num_related
+        self.max_supported = max_supported
+
+
+class ModelBackend:
+    def __init__(self, local_unit_name):
+        local_app_name = local_unit_name.split('/')[0]
+        self.local_app = Application(local_app_name)
+        self.local_unit = Unit(local_unit_name, self.local_app)
```

Comment:
I suggest drawing the line so that the backend provides data, not instances, not caching. Otherwise we're moving the business logic of how to implement the model into the backend, and it's becoming more of a frontend than a backend -- for every backend that we implement we'll need to redo all of that.